external-crates/move/documentation/book/src/method-syntax.md

```diff
@@ -0,0 +1,442 @@
+}
+
+module b::other {
+    use a::m::X;
+
+    use fun a::m::foo as X.f;
+
+    fun example(x: &X) {
+        x.f(); // resolves to a::m::foo
+        {
+            use a::m::bar as f;
+            x.f(); // resolves to a::m::bar
+        };
+        x.f(); // still resolves to a::m::foo
+        {
+            use fun a::m::bar as X.f;
+            x.f(); // resolves to a::m::bar
+        }
+    }
+```
+
+## Automatic Borrowing
+
+When resolving a method, the compiler will automatically borrow the receiver if the function expects
+a reference. For example
+
+```move
+module a::m {
+    public struct X() has copy, drop;
+    public fun by_val(_: X) {}
+    public fun by_ref(_: &X) {}
+    public fun by_mut(_: &mut X) {}
+
+    fun example(mut x: X) {
+        x.by_ref(); // resolves to a::m::by_ref(&x)
+        x.by_mut(); // resolves to a::m::by_mut(&mut x)
+    }
+}
+```
+
+In these examples, `x` was automatically borrowed to `&x` and `&mut x` respectively. This will also
+work through field access
+
+```move
+module a::m {
+    public struct X() has copy, drop;
+    public fun by_val(_: X) {}
+    public fun by_ref(_: &X) {}
+    public fun by_mut(_: &mut X) {}
+
+    public struct Y has drop { x: X }
+
+    fun example(mut y: Y) {
+        y.x.by_ref(); // resolves to a::m::by_ref(&y.x)
+        y.x.by_mut(); // resolves to a::m::by_mut(&mut y.x)
+    }
+}
+```
+
+Note that in both examples, the local variable had to be labeled as [`mut`](./variables.md) to allow
+for the `&mut` borrow. Without this, there would be an error saying that `x` (or `y` in the second
+example) is not mutable.
+
+Keep in mind that without a reference, normal rules for variable and field access come into play.
+Meaning a value might be moved or copied if it is not borrowed.
+
+```move
+module a::m {
+    public struct X() has copy, drop;
+    public fun by_val(_: X) {}
+    public fun by_ref(_: &X) {}
+    public fun by_mut(_: &mut X) {}
+
+    public struct Y has drop { x: X }
+    public fun drop_y(y: Y) { y }
+
+    fun example(y: Y) {
+        y.x.by_val(); // copies `y.x` since `by_val` is by-value and `X` has `copy`
+        y.drop_y(); // moves `y` since `drop_y` is by-value and `Y` does _not_ have `copy`
+    }
+}
+```
+
+## Chaining
+
+Method calls can be chained, since the receiver of the method can be any expression.
+
+```move
+module a::shapes {
+    public struct Point has copy, drop, store { x: u64, y: u64 }
+    public struct Line has copy, drop, store { start: Point, end: Point }
+
+    public fun x(p: &Point): u64 { p.x }
+    public fun y(p: &Point): u64 { p.y }
+
+    public fun start(l: &Line): &Point { &l.start }
+    public fun end(l: &Line): &Point { &l.end }
+
+}
+
+module b::example {
+    use a::shapes::Line;
+
+    public fun x_values(l: Line): (u64, u64) {
+        (l.start().x(), l.end().x())
+    }
+
+}
+```
+
+In this example, `l.start().x()` the compiler first resovles `l.start()` to `a::shapes::start(&l)`.
```

Comment:
```suggestion
In this example, `l.start().x()` the compiler first resolves `l.start()` to `a::shapes::start(&l)`.
```
I think this sentence is missing a word or three somewhere around `l.start().x()`?